src/main/kotlin/org/rust/ide/formatter/RustSpacing.kt

```diff
@@ -1,124 +1,198 @@
+                         origNode2: ASTNode, origPsi2: PsiElement): Pair<PsiElement, PsiElement> =
+    Pair(
+        if (origPsi1 is PsiComment && origNode1.hasLineBreakAfterInSameParent) {
+            origPsi1.getPrevNonCommentSibling() ?: origPsi1
+        } else {
+            origPsi1
+        },
+        if (origPsi2 is PsiComment && origNode2.hasLineBreakBeforeInSameParent) {
+            origPsi2.getNextNonCommentSibling() ?: origPsi2
+        } else {
+            origPsi2
+        }
+    )
+
+private val KEYWORDS = ts(*IElementType.enumerate { it is RustKeywordTokenType })
+private val NO_SPACE_AROUND_OPS = ts(COLONCOLON, DOT, DOTDOT)
+private val SPACE_AROUND_OPS = ts(AND, ANDAND, ANDEQ, ARROW, FAT_ARROW, DIV, DIVEQ, EQ, EQEQ,
+    EXCLEQ, GT, LT, MINUSEQ, MUL, MULEQ, OR, OREQ, OROR, PLUSEQ, REM, REMEQ, XOR, XOREQ, MINUS, PLUS,
+    GTGTEQ, GTGT, GTEQ, LTLTEQ, LTLT, LTEQ)
+private val UNARY_OPS = ts(MINUS, MUL, EXCL, AND, ANDAND)
+private val SPACE_AFTER_OPS = ts(COMMA, COLON, SEMICOLON)
+// PATH_PART because `Fn(A) -> R`
+private val PAREN_LIST_HOLDERS = ts(PAREN_EXPR, TUPLE_EXPR, TUPLE_TYPE, PARAMETERS, VARIADIC_PARAMETERS, ARG_LIST,
+    IMPL_METHOD_MEMBER, BARE_FN_TYPE, PATH_PART, PAT_ENUM, PAT_TUP, ENUM_TUPLE_ARGS)
+private val BRACK_LIST_HOLDERS = ts(VEC_TYPE, ARRAY_EXPR, INDEX_EXPR)
+private val BRACE_LIST_HOLDERS = ts(USE_GLOB_LIST)
+private val ANGLE_LIST_HOLDERS = ts(GENERIC_PARAMS, GENERIC_ARGS, QUAL_PATH_EXPR)
+private val ATTRS = ts(OUTER_ATTR, INNER_ATTR)
+private val BLOCK_LIKE = ts(BLOCK, STRUCT_DECL_ARGS, STRUCT_EXPR_BODY, IMPL_BODY, MATCH_BODY, TRAIT_BODY, ENUM_BODY,
+    ENUM_STRUCT_ARGS)
+private val TYPES = ts(VEC_TYPE, PTR_TYPE, REF_TYPE, BARE_FN_TYPE, TUPLE_TYPE, PATH_TYPE,
+    TYPE_WITH_BOUNDS_TYPE, FOR_IN_TYPE, WILDCARD_TYPE)
+private val MACRO_ARGS = ts(MACRO_ARG, FORMAT_MACRO_ARGS, TRY_MACRO_ARGS)
+private val PARAMS_LIKE = ts(PARAMETERS, VARIADIC_PARAMETERS)
+
+private inline fun SpacingBuilder.applyForEach(
+    tokenSet: TokenSet, block: SpacingBuilder.(IElementType) -> SpacingBuilder): SpacingBuilder {
+    var self = this
+    for (tt in tokenSet.types) {
+        self = block(this, tt)
+    }
+    return self
+}
+
+private fun lineBreak(minLineFeeds: Int = 1,
+                      keepLineBreaks: Boolean = true,
+                      keepBlankLines: Int = 1): Spacing =
+    Spacing.createSpacing(0, Int.MAX_VALUE, minLineFeeds, keepLineBreaks, keepBlankLines)
+
+private val PsiElement.isTopLevelItem: Boolean
+    get() = (this is RustItem || this is RustAttr) && this.parent is RustFile
+
+private val PsiElement.isStmtOrExpr: Boolean
+    get() = this is RustStmt || this is RustExpr
+
+private val ASTNode.hasLineBreakAfterInSameParent: Boolean
+    get() = treeNext != null && isWhiteSpaceWithLineBreak(TreeUtil.findFirstLeaf(treeNext))
+
+private val ASTNode.hasLineBreakBeforeInSameParent: Boolean
+    get() = treePrev != null && isWhiteSpaceWithLineBreak(TreeUtil.findLastLeaf(treePrev))
```

Comment:
I'm not sure whether `TreeUtil` is available in non-idea ides; though these methods are almost copy-pasted from Python's formatter; so it has to be available at least in PyCharm.